scikeras/_utils.py

```diff
@@ -231,8 +240,113 @@ def has_param(func: Callable, param: str) -> bool:
 
 
 def accepts_kwargs(func: Callable) -> bool:
+    """Check if ``func`` accepts kwargs.
+    """
     return any(
         True
         for param in inspect.signature(func).parameters.values()
         if param.kind == param.VAR_KEYWORD
     )
+
+
+def compile_with_params(items, params, base_params=None):
+    """Recursively compile nested structures of classes
+    using parameters from params.
+    """
+    if isclass(items):
+        item = items
+        new_base_params = {p: v for p, v in params.items() if "__" not in p}
+        base_params = base_params or dict()
+        kwargs = {**base_params, **new_base_params}
+        for p, v in kwargs.items():
+            kwargs[p] = compile_with_params(
+                items=v,
+                params=route_params(
+                    params=params,
+                    destination=f"{p}",
+                    pass_filter=set(),
+                    strict=False,
+                ),
+            )
+        return item(**kwargs)
+    if isinstance(items, (list, tuple)):
+        iter_type_ = type(items)
+        res = list()
+        new_base_params = {p: v for p, v in params.items() if "__" not in p}
+        for idx, item in enumerate(items):
+            item_params = route_params(
+                params=params,
+                destination=f"{idx}",
+                pass_filter=set(),
+                strict=False,
+            )
+            res.append(
+                compile_with_params(
+                    items=item, params=item_params, base_params=new_base_params
+                )
+            )
+        return iter_type_(res)
+    if isinstance(items, (dict,)):
+        res = dict()
+        new_base_params = {p: v for p, v in params.items() if "__" not in p}
+        for key, item in items.items():
+            item_params = route_params(
+                params=params,
+                destination=f"{key}",
+                pass_filter=set(),
+                strict=False,
+            )
+            res[key] = compile_with_params(
+                items=item, params=item_params, base_params=new_base_params,
+            )
+        return res
+    # non-compilable item, check if it has any routed parameters
+    item = items
+    new_base_params = {p: v for p, v in params.items() if "__" not in p}
+    base_params = base_params or dict()
+    kwargs = {**base_params, **new_base_params}
+    if kwargs:
+        raise TypeError(
+            f'TypeError: "{str(item)}" object of type "{type(item)}"'
+            "does not accept parameters because it's not a class."
+            f' However, it received parameters "{kwargs}"'
+        )
+    return item
+
+
+def _class_from_strings(items, item_type: str):
```

Comment:
Happy to change it if it makes things easier, but could you help me understand why this makes it easier?

`item/items` is an iterable/dict or single item depending on the recursion level. Is there any convention for singular vs. plural in this case?

And why the default value for `type`? It is and should be always called with a value.

Reply:
How 'bout this? 

``` python
def _class_from_string(items, type=object):
```

Even cleaner would be

``` python
def _class_from_string(items: Union[str, dict, tuple, list], classes=tf.keras.losses.get):
```

> but could you help me understand why this makes it easier?

I ran into this code:

https://github.com/adriangb/scikeras/blob/471ca9f74db94a732030d33b80298996adbd0843/scikeras/wrappers.py#L260

and had to keep on scrolling back and forth to see what `"optimizer"` was and what exactly the function returned.

Reply:
> def _class_from_string(items: Union[str, dict, tuple, list], classes=tf.keras.losses.get):

Yes! I was actually thinking of doing something like this (passing the function instead of the string) the other day when I was offline, but I forgot about this idea. Thank you for suggesting it!

Reply:
I was able to remove all of those `if` statements as well. Looks much better now!